src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -273,4 +370,74 @@ async Task LogDiagnosticsAsync(ImmutableArray<DiagnosticLogItem> diagnosticLogIt
             await toastErrorReporter.ReportErrorAsync(worstLspMessageKind, message, cancellationToken);
         }
     }
+
+    protected async ValueTask BeginLoadingProjectAsync(string projectPath, ProjectId primordialProjectId)
+    {
+        using (await _gate.DisposableWaitAsync(CancellationToken.None))
+        {
+            if (_loadedProjects.ContainsKey(projectPath))
+            {
+                return;
+            }
+
+            _loadedProjects.Add(projectPath, new ProjectLoadState.BeginLoadingWithPrimordial(primordialProjectId));
+            _projectsToLoadAndReload.AddWork(new ProjectToLoad(projectPath, ProjectGuid: null, ReportTelemetry: true));
+        }
+    }
+
+    protected async Task LoadProjectsAsync(ImmutableArray<(string ProjectPath, string? ProjectGuid)> projectPaths, CancellationToken cancellationToken)
+    {
+        using (await _gate.DisposableWaitAsync(cancellationToken))
+        {
+            foreach (var (path, guid) in projectPaths)
+            {
+                if (_loadedProjects.ContainsKey(path))
+                {
+                    continue;
+                }
+
+                _loadedProjects.Add(path, new ProjectLoadState.BeginLoading());
+                _projectsToLoadAndReload.AddWork(new ProjectToLoad(Path: path, ProjectGuid: guid, ReportTelemetry: true));
+            }
+
+            // TODO: since this method is used to implement 'OpenSolution' and 'OpenProjects',
+            // it seems like currently loaded projects which are not included in 'projectPaths' should be unloaded.
+        }
+
+        await _projectsToLoadAndReload.WaitUntilCurrentBatchCompletesAsync();
```

Comment:
wary about this.  does some caller actually need to wait on this work being done.  ideally the work can be done and can finish at its own cadence and the caller doesn't care.

Reply:
can we doc this?  what is this "projects finished loading" notification?  i'm not sure i'm familliar with this, or what we need it for.

Reply:
It looks like the primary use for it is to allow integration tests to wait for it before performing some actions.

Reply:
yup as Rikki mentioned, integration tests is one case.   Our client also uses it to trigger a toast to suggest launch.json/tasks.json generation, trigger refresh of active project contexts, etc.

Reply:
We also do this for things like integration test reliability: it means that the "open solution" command send from VS Code doesn't complete until the solution is loaded.

That said, that really should be restricted to the command implementations for doing this. I'd be a bit worried that somebody might accidentally await this where they don't realize the ramifications of this. I'd say maybe consider having a separate protected method which just calls the WaitForCurrentBatchCompletes and that's used for open solution/project specifically.

Reply:
note: integration tests should generally be using the IAsyncOpListener.  That's waht it is there for, and why we thread it into the ABWQ.  

Note that we generaly are loath to await this stuff, because effectively we're saying: add to a queue, with artiffiial delays... but then have something wait on that work.

Reply:
my concern here is non-blocking (no pun intended).  just raising it as something that i feel smells, and would prefer we avoid.  but that can happen later :)

Reply:
> note: integration tests should generally be using the IAsyncOpListener.

It's a bit murkier once there's cross-process stuff going on with VS Code and the LSP queue too. It also just means any other client code in another extension invoking the VS Code command knows when it's done too.

The wait should exist...at the caller. Not here.